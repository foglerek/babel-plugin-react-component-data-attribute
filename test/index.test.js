import {resolve} from 'path';
import * as babel from '@babel/core';
import babelPluginReactComponentDataAttribute from '../src';

describe('babelPluginReactComponentDataAttribute()', () => {
  it('handles simple returns', () => {
    expect(transform(`
      function MyComponent() {
        return <div />;
      }
    `)).toMatchSnapshot();
  });

  it('does not add attributes to composite components', () => {
    expect(transform(`
      function MyComponent() {
        return <SomeOtherComponent />;
      }
    `)).toMatchSnapshot();
  });

  it('adds the attribute to the first non-composite component', () => {
    expect(transform(`
      function MyComponent() {
        return <SomeOtherComponent><div><div /></div></SomeOtherComponent>;
      }
    `)).toMatchSnapshot();

    expect(transform(`
      function MyComponent() {
        return (
          <SomeOtherComponent>{() => <div><div /></div>}</SomeOtherComponent>
        );
      }
    `)).toMatchSnapshot();
  });

  it('does not add attributes to JSX elements in props', () => {
    expect(transform(`
      function MyComponent() {
        return <SomeOtherComponent someProp={<div />}><div /></SomeOtherComponent>;
      }
    `)).toMatchSnapshot();
  });

  it('handles conditional returns', () => {
    expect(transform(`
      function MyComponent() {
        return true
          ? <div />
          : <span />;
      }
    `)).toMatchSnapshot();

    expect(transform(`
      function MyComponent() {
        if (true) {
          return <div />;
        } else {
          return <span />;
        }
      }
    `)).toMatchSnapshot();
  });

  it('only adds the attribute to the top-level node', () => {
    expect(transform(`
      function MyComponent() {
        return (
          <div><div /></div>
        );
      }
    `)).toMatchSnapshot();
  });

  it('does not add attributes from a parent function', () => {
    expect(transform(`
      function MyComponent() {
        return () => <div />;
      }
    `)).toMatchSnapshot();
  });

  it('adds the property to React.createElement calls', () => {
    expect(transform(`
      function MyComponent() {
        return React.createElement(someElement, {foo: 'bar'});
      }
    `)).toMatchSnapshot();

    expect(transform(`
      function MyComponent() {
        return React.createElement(someElement);
      }
    `)).toMatchSnapshot();
  });

  it('updates intermediate assignments as appropriate', () => {
    expect(transform(`
      function MyComponent() {
        const markup = <div />;
        return markup;
      }
    `)).toMatchSnapshot();

    expect(transform(`
      function MyComponent() {
        const markup = <SomeOtherComponent />;
        return markup;
      }
    `)).toMatchSnapshot();

    expect(transform(`
      function MyComponent() {
        const markup = null;
        return markup;
      }
    `)).toMatchSnapshot();

    expect(transform(`
      function MyComponent() {
        const markup = React.createElement('div', {});
        return markup;
      }
    `)).toMatchSnapshot();
  });

  it('only processes top-level components', () => {
    expect(transform(`
      if (true) {
        function MyComponent() {
          return <div />;
        }
      }
    `)).toMatchSnapshot();

    expect(transform(`
      function MyComponent() {}

      MyComponent.prototype.render = function render() {
        return <div />;
      }
    `)).toMatchSnapshot();
  });

  it('does not process nodes that already have the data attribute', () => {
    expect(transform(`
      function MyComponent() {
        return <div data-component="MyComponent" />;
      }
    `)).toMatchSnapshot();

    expect(transform(`
      function MyComponent() {
        return React.createElement('div', {'data-component': 'MyComponent'});
      }
    `)).toMatchSnapshot();
  });

  describe('name', () => {
    it('uses the variable name when no name exists', () => {
      expect(transform(`
        const MyComponent = class extends React.Component {
          render() {
            return <div />;
          }
        }
      `)).toMatchSnapshot();

      expect(transform(`
        const MyComponent = function() {
          return <div />;
        }
      `)).toMatchSnapshot();
    });

    it('uses the file’s basename when it is not unknown and is not an index file', () => {
      const filename = resolve('MyComponent.js');

      expect(transform(`
        export default class extends React.Component {
          render() {
            return <div />;
          }
        }
      `, {}, {filename})).toMatchSnapshot();

      expect(transform(`
        export default function() {
          return <div />;
        }
      `, {}, {filename})).toMatchSnapshot();

      expect(transform(`
        export default () => {
          return <div />;
        }
      `, {}, {filename})).toMatchSnapshot();
    });

    it('uses the file’s directory name when it is an index file', () => {
      const filename = resolve('MyComponent/index.js');

      expect(transform(`
        export default class extends React.Component {
          render() {
            return <div />;
          }
        }
      `, {}, {filename})).toMatchSnapshot();

      expect(transform(`
        export default function() {
          return <div />;
        }
      `, {}, {filename})).toMatchSnapshot();

      expect(transform(`
        export default () => {
          return <div />;
        }
      `, {}, {filename})).toMatchSnapshot();
    });
  });

  describe('arrow expressions', () => {
    it('handles non-block statement arrow expressions', () => {
      expect(transform(`
        const MyComponent = () => <div />;
      `)).toMatchSnapshot();
    });
  });

  describe('classes', () => {
    it('only adds attributes to render methods', () => {
      expect(transform(`
        class MyComponent extends React.Component {
          renderAnotherThing() {
            return <div />;
          }
        }
      `)).toMatchSnapshot();
    });
  });

  describe('options', () => {
    describe('onlyRootComponents', () => {
      it('adds attributes to components that are "root" ones', () => {
        expect(transform(`
          export default function MyComponent() {
            return <div />;
          }
        `, {onlyRootComponents: true}, {filename: resolve('MyComponent/index.js')})).toMatchSnapshot();

        expect(transform(`
          export default function SomeComponent() {
            return <div />;
          }
        `, {}, {filename: resolve('MyComponent/MyComponent.js')})).toMatchSnapshot();

        expect(transform(`
          export function MyComponent() {
            return <div />;
          }
        `, {onlyRootComponents: true}, {filename: resolve('MyComponent/MyComponent.js')})).toMatchSnapshot();

        expect(transform(`
          function MyComponent() {
            return <div />;
          }

          export default MyComponent;
        `, {onlyRootComponents: true}, {filename: resolve('MyComponent/MyComponent.js')})).toMatchSnapshot();

        expect(transform(`
          function MyComponent() {
            return <div />;
          }

          export default doSomethingTo(MyComponent);
        `, {onlyRootComponents: true}, {filename: resolve('MyComponent/MyComponent.js')})).toMatchSnapshot();

        expect(transform(`
          function MyComponent() {
            return <div />;
          }

          export {MyComponent};
        `, {onlyRootComponents: true}, {filename: resolve('MyComponent/MyComponent.js')})).toMatchSnapshot();
      });

      it('does not add data attributes to other components', () => {
        expect(transform(`
          function SomeComponent() {
            return <div />;
          }
        `, {onlyRootComponents: true}, {filename: resolve('MyComponent/index.js')})).toMatchSnapshot();

        expect(transform(`
          export default function SomeComponent() {
            return <div />;
          }
        `, {onlyRootComponents: true}, {filename: resolve('MyComponent/SomethingElse.js')})).toMatchSnapshot();
      });
    });

    describe('overrides', () => {
      describe('process', () => {
        it('does not process a component when it would have otherwise', () => {
          expect(transform(`
            export default function MyComponent() {
              return <div />;
            }
          `, {
            onlyRootComponents: true,
            overrides: {
              MyComponent: {process: false},
            },
          }, {
            filename: resolve('MyComponent/index.js'),
          })).toMatchSnapshot();
        });

        it('processes a component when it would not have otherwise', () => {
          expect(transform(`
            export default function MyNestedComponent() {
              return <div />;
            }
          `, {
            onlyRootComponents: true,
            overrides: {
              MyNestedComponent: {process: true},
            },
          }, {
            filename: resolve('MyComponent/MyNestedComponent.js'),
          })).toMatchSnapshot();
        });
      });

      describe('methods', () => {
        it('processes return statements in the specified methods', () => {
          expect(transform(`
            export default class MyComponent extends React.Component {
              renderLayer() {
                return <div />;
              }

              renderSomethingElse() {
                return <div />;
              }

              render() {
                return <div />;
              }
            }
          `, {
            overrides: {
              MyComponent: {
                methods: ['renderLayer', 'renderSomethingElse'],
              },
            },
          })).toMatchSnapshot();
        });
      });

      describe('name', () => {
        it('uses the custom name', () => {
          expect(transform(`
            class MyComponent extends React.Component {
              render() {
                return <div />;
              }
            }
          `, {
            overrides: {
              MyComponent: {
                name: 'SomeOtherComponent',
              },
            },
          })).toMatchSnapshot();
        });
      });
    });
  });
});

function transform(code, pluginOptions, transformOptions) {
  return babel.transform(code, {
    babelrc: false,
    plugins: [
      [babelPluginReactComponentDataAttribute, pluginOptions],
    ],
    parserOpts: {
      plugins: ['jsx'],
    },
    ...transformOptions,
  }).code.trim();
}
