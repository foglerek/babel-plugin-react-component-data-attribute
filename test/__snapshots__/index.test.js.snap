// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babelPluginReactComponentDataAttribute() adds the attribute to the first non-composite component 1`] = `
"function MyComponent() {
  return <SomeOtherComponent><div data-component=\\"MyComponent\\"><div /></div></SomeOtherComponent>;
}"
`;

exports[`babelPluginReactComponentDataAttribute() adds the attribute to the first non-composite component 2`] = `
"function MyComponent() {
  return <SomeOtherComponent>{() => <div><div /></div>}</SomeOtherComponent>;
}"
`;

exports[`babelPluginReactComponentDataAttribute() adds the property to React.createElement calls 1`] = `
"function MyComponent() {
  return React.createElement(someElement, {
    foo: 'bar',
    \\"data-component\\": \\"MyComponent\\"
  });
}"
`;

exports[`babelPluginReactComponentDataAttribute() adds the property to React.createElement calls 2`] = `
"function MyComponent() {
  return React.createElement(someElement, {
    \\"data-component\\": \\"MyComponent\\"
  });
}"
`;

exports[`babelPluginReactComponentDataAttribute() arrow expressions handles non-block statement arrow expressions 1`] = `"const MyComponent = () => <div data-component=\\"MyComponent\\" />;"`;

exports[`babelPluginReactComponentDataAttribute() classes only adds attributes to render methods 1`] = `
"class MyComponent extends React.Component {
  renderAnotherThing() {
    return <div />;
  }

}"
`;

exports[`babelPluginReactComponentDataAttribute() does not add attributes from a parent function 1`] = `
"function MyComponent() {
  return () => <div />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() does not add attributes to JSX elements in props 1`] = `
"function MyComponent() {
  return <SomeOtherComponent someProp={<div />}><div data-component=\\"MyComponent\\" /></SomeOtherComponent>;
}"
`;

exports[`babelPluginReactComponentDataAttribute() does not add attributes to composite components 1`] = `
"function MyComponent() {
  return <SomeOtherComponent />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() does not process nodes that already have the data attribute 1`] = `
"function MyComponent() {
  return <div data-component=\\"MyComponent\\" />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() does not process nodes that already have the data attribute 2`] = `
"function MyComponent() {
  return React.createElement('div', {
    'data-component': 'MyComponent'
  });
}"
`;

exports[`babelPluginReactComponentDataAttribute() handles conditional returns 1`] = `
"function MyComponent() {
  return true ? <div data-component=\\"MyComponent\\" /> : <span data-component=\\"MyComponent\\" />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() handles conditional returns 2`] = `
"function MyComponent() {
  if (true) {
    return <div data-component=\\"MyComponent\\" />;
  } else {
    return <span data-component=\\"MyComponent\\" />;
  }
}"
`;

exports[`babelPluginReactComponentDataAttribute() handles simple returns 1`] = `
"function MyComponent() {
  return <div data-component=\\"MyComponent\\" />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() name uses the file’s basename when it is not unknown and is not an index file 1`] = `
"export default class extends React.Component {
  render() {
    return <div data-component=\\"MyComponent\\" />;
  }

}"
`;

exports[`babelPluginReactComponentDataAttribute() name uses the file’s basename when it is not unknown and is not an index file 2`] = `
"export default function () {
  return <div data-component=\\"MyComponent\\" />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() name uses the file’s basename when it is not unknown and is not an index file 3`] = `
"export default (() => {
  return <div data-component=\\"MyComponent\\" />;
});"
`;

exports[`babelPluginReactComponentDataAttribute() name uses the file’s directory name when it is an index file 1`] = `
"export default class extends React.Component {
  render() {
    return <div data-component=\\"MyComponent\\" />;
  }

}"
`;

exports[`babelPluginReactComponentDataAttribute() name uses the file’s directory name when it is an index file 2`] = `
"export default function () {
  return <div data-component=\\"MyComponent\\" />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() name uses the file’s directory name when it is an index file 3`] = `
"export default (() => {
  return <div data-component=\\"MyComponent\\" />;
});"
`;

exports[`babelPluginReactComponentDataAttribute() name uses the variable name when no name exists 1`] = `
"const MyComponent = class extends React.Component {
  render() {
    return <div data-component=\\"MyComponent\\" />;
  }

};"
`;

exports[`babelPluginReactComponentDataAttribute() name uses the variable name when no name exists 2`] = `
"const MyComponent = function () {
  return <div data-component=\\"MyComponent\\" />;
};"
`;

exports[`babelPluginReactComponentDataAttribute() only adds the attribute to the top-level node 1`] = `
"function MyComponent() {
  return <div data-component=\\"MyComponent\\"><div /></div>;
}"
`;

exports[`babelPluginReactComponentDataAttribute() only processes top-level components 1`] = `
"if (true) {
  function MyComponent() {
    return <div data-component=\\"MyComponent\\" />;
  }
}"
`;

exports[`babelPluginReactComponentDataAttribute() only processes top-level components 2`] = `
"function MyComponent() {}

MyComponent.prototype.render = function render() {
  return <div />;
};"
`;

exports[`babelPluginReactComponentDataAttribute() options onlyRootComponents adds attributes to components that are "root" ones 1`] = `
"export default function MyComponent() {
  return <div data-component=\\"MyComponent\\" />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() options onlyRootComponents adds attributes to components that are "root" ones 2`] = `
"export default function SomeComponent() {
  return <div data-component=\\"SomeComponent\\" />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() options onlyRootComponents adds attributes to components that are "root" ones 3`] = `
"export function MyComponent() {
  return <div data-component=\\"MyComponent\\" />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() options onlyRootComponents adds attributes to components that are "root" ones 4`] = `
"function MyComponent() {
  return <div data-component=\\"MyComponent\\" />;
}

export default MyComponent;"
`;

exports[`babelPluginReactComponentDataAttribute() options onlyRootComponents adds attributes to components that are "root" ones 5`] = `
"function MyComponent() {
  return <div data-component=\\"MyComponent\\" />;
}

export default doSomethingTo(MyComponent);"
`;

exports[`babelPluginReactComponentDataAttribute() options onlyRootComponents adds attributes to components that are "root" ones 6`] = `
"function MyComponent() {
  return <div data-component=\\"MyComponent\\" />;
}

export { MyComponent };"
`;

exports[`babelPluginReactComponentDataAttribute() options onlyRootComponents does not add data attributes to other components 1`] = `
"function SomeComponent() {
  return <div />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() options onlyRootComponents does not add data attributes to other components 2`] = `
"export default function SomeComponent() {
  return <div />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() options overrides methods processes return statements in the specified methods 1`] = `
"export default class MyComponent extends React.Component {
  renderLayer() {
    return <div data-component=\\"MyComponent\\" />;
  }

  renderSomethingElse() {
    return <div data-component=\\"MyComponent\\" />;
  }

  render() {
    return <div />;
  }

}"
`;

exports[`babelPluginReactComponentDataAttribute() options overrides name uses the custom name 1`] = `
"class MyComponent extends React.Component {
  render() {
    return <div data-component=\\"SomeOtherComponent\\" />;
  }

}"
`;

exports[`babelPluginReactComponentDataAttribute() options overrides process does not process a component when it would have otherwise 1`] = `
"export default function MyComponent() {
  return <div />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() options overrides process processes a component when it would not have otherwise 1`] = `
"export default function MyNestedComponent() {
  return <div data-component=\\"MyNestedComponent\\" />;
}"
`;

exports[`babelPluginReactComponentDataAttribute() updates intermediate assignments as appropriate 1`] = `
"function MyComponent() {
  const markup = <div data-component=\\"MyComponent\\" />;
  return markup;
}"
`;

exports[`babelPluginReactComponentDataAttribute() updates intermediate assignments as appropriate 2`] = `
"function MyComponent() {
  const markup = <SomeOtherComponent />;
  return markup;
}"
`;

exports[`babelPluginReactComponentDataAttribute() updates intermediate assignments as appropriate 3`] = `
"function MyComponent() {
  const markup = null;
  return markup;
}"
`;

exports[`babelPluginReactComponentDataAttribute() updates intermediate assignments as appropriate 4`] = `
"function MyComponent() {
  const markup = React.createElement('div', {
    \\"data-component\\": \\"MyComponent\\"
  });
  return markup;
}"
`;
